provider/pkg/metadata/naming.go

```diff
@@ -15,22 +15,36 @@
 package metadata
 
 import (
+	"errors"
+
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
 // AssignNameIfAutonamable generates a name for an object. Uses DNS-1123-compliant characters.
 // All auto-named resources get the annotation `pulumi.com/autonamed` for tooling purposes.
-func AssignNameIfAutonamable(randomSeed []byte, obj *unstructured.Unstructured, propMap resource.PropertyMap, urn resource.URN) {
+func AssignNameIfAutonamable(randomSeed []byte, engineAutonaming *pulumirpc.CheckRequest_AutonamingOptions,
+	obj *unstructured.Unstructured, propMap resource.PropertyMap, urn resource.URN) error {
 	contract.Assertf(urn.Name() != "", "expected non-empty name in URN: %s", urn)
-	if !IsNamed(obj, propMap) && !IsGenerateName(obj, propMap) {
-		prefix := urn.Name() + "-"
-		autoname, err := resource.NewUniqueName(randomSeed, prefix, 0, 0, nil)
-		contract.AssertNoErrorf(err, "unexpected error while creating NewUniqueName")
-		obj.SetName(autoname)
-		SetAnnotationTrue(obj, AnnotationAutonamed)
+	if IsNamed(obj, propMap) || IsGenerateName(obj, propMap) {
+		return nil
+	}
+	prefix := urn.Name() + "-"
+	autoname, err := resource.NewUniqueName(randomSeed, prefix, 0, 0, nil)
+	contract.AssertNoErrorf(err, "unexpected error while creating NewUniqueName")
+	if engineAutonaming != nil {
+		switch engineAutonaming.Mode {
+		case pulumirpc.CheckRequest_AutonamingOptions_DISABLE:
+			return errors.New("autonaming is disabled, please provide a name or generateName value")
```

Comment:
Note that other error messages within the k provider use a slightly different wording, for example:
> "patch resources require the `.metadata.name` field to be set"

And, why is this different than azure-native w.r.t. to returning an error in the 'disabled' case?
https://github.com/pulumi/pulumi-azure-native/blob/b5b0a217b37ebe605bc871b552d3dc195e882d22/provider/pkg/provider/provider.go#L570-L573

Reply:
> Note that other error messages within the k provider use a slightly different wording

Great point. Should it be `autonaming is disabled, resources require the .metadata.name field to be set`?

> why is this different than azure-native w.r.t. to returning an error in the 'disabled' case?

The end user UX is the same - users will get an error from Check. It's more of an implementation detail. In Azure Native, the `getDefaultName` doesn't return an `error`, but it's produced two levels up in Check itself, where it validates required properties. That was already the case before autonaming configuration.